packages/flutter_tools/lib/src/build_system/targets/web.dart

```diff
@@ -393,40 +402,101 @@ class WebServiceWorker extends Target {
 /// Generate a service worker with an app-specific cache name a map of
 /// resource files.
 ///
-/// We embed file hashes directly into the worker so that the byte for byte
+/// The tool embeds file hashes directly into the worker so that the byte for byte
 /// invalidation will automatically reactivate workers whenever a new
 /// version is deployed.
-// TODO(jonahwilliams): on re-activate, only evict stale assets.
-String generateServiceWorker(Map<String, String> resources) {
+String generateServiceWorker(Map<String, String> resources, List<String> coreBundle) {
   return '''
 'use strict';
+const MANIFEST = 'flutter-app-manifest';
+const TEMP = 'flutter-temp-cache';
 const CACHE_NAME = 'flutter-app-cache';
 const RESOURCES = {
   ${resources.entries.map((MapEntry<String, String> entry) => '"${entry.key}": "${entry.value}"').join(",\n")}
 };
 
-self.addEventListener('activate', function (event) {
-  event.waitUntil(
-    caches.keys().then(function (cacheName) {
-      return caches.delete(cacheName);
-    }).then(function (_) {
-      return caches.open(CACHE_NAME);
-    }).then(function (cache) {
-      return cache.addAll(Object.keys(RESOURCES));
+const CORE = [
+  ${coreBundle.map((String file) => '"$file"').join(',\n')}];
+
+self.addEventListener("install", (event) => {
+  return event.waitUntil(
+    caches.open(TEMP).then((cache) => {
+      return cache.addAll(CORE);
     })
   );
 });
 
-self.addEventListener('fetch', function (event) {
-  event.respondWith(
-    caches.match(event.request)
-      .then(function (response) {
-        if (response) {
-          return response;
+
+self.addEventListener("install", (event) => {
+  return event.waitUntil(
```

Comment:
During install, we unconditionally cache the minimal set of assets needed to run:

main.dart.js, index.html, AssetManifest.json, FontManifest.json, LICENSE

Reply:
`main.dart.js` depends on the name you choose for `lib/main.dart`, doesn't it? Do we need to document that you need to call your main entrypoint `main.dart` for it to be cacheable by the service worker?

Reply:
No, the output is always named `main.dart.js`, see also:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_system/targets/web.dart#L164